### PR TITLE
[SPARK-27074][SQL][test-hadoop3.2][test-maven] Hive 3.1 metastore support HiveClientImpl.runHive

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

Hive 3.1.1's `CommandProcessor` have 2 changes:
1. [HIVE-17626](https://issues.apache.org/jira/browse/HIVE-17626)(Hive 3.0.0) add ReExecDriver. So the current code path is: https://github.com/apache/spark/blob/02bbe977abaf7006b845a7e99d612b0235aa0025/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L736-L742
We can disable `hive.query.reexecution.enabled` to workaround this change.
2. [HIVE-18238](http://issues.apache.org/jira/browse/HIVE-18238)(Hive 3.0.0) changed the `Driver.close()` function return type. We can workaround it by ` driver.getClass.getMethod("close").invoke(driver)`

So Hive 3.1 metastore could support `HiveClientImpl.runHive` after this pr.

## How was this patch tested?

unit tests

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 1]
```diff
@@ -711,14 +711,26 @@ private[hive] class HiveClientImpl(
...
       val cmd_trimmed: String = cmd.trim()
       val tokens: Array[String] = cmd_trimmed.split("\\s+")
       // The remainder of the command.
       val cmd_1: String = cmd_trimmed.substring(tokens(0).length()).trim()
       val proc = shim.getCommandProcessor(tokens(0), conf)
       proc match {
-        case driver: Driver =>
+        // HIVE-18238(Hive 3.0.0) changed the close() function return type.
+        // This change is not compatible with the built-in Hive.
+        case driver: Driver if version != hive.v3_1 =>
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Error message:
```java
[info] org.apache.spark.sql.hive.client.VersionsSuite *** ABORTED *** (5 minutes, 8 seconds)
[info]   java.lang.NoSuchMethodError: org.apache.hadoop.hive.ql.Driver.close()I
[info]   at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$runHive$1(HiveClientImpl.scala:744)
[info]   at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$withHiveState$1(HiveClientImpl.scala:277)
[info]   at org.apache.spark.sql.hive.client.HiveClientImpl.liftedTree1$1(HiveClientImpl.scala:215)
[info]   at org.apache.spark.sql.hive.client.HiveClientImpl.retryLocked(HiveClientImpl.scala:214)
[info]   at org.apache.spark.sql.hive.client.HiveClientImpl.withHiveState(HiveClientImpl.scala:260)
[info]   at org.apache.spark.sql.hive.client.HiveClientImpl.runHive(HiveClientImpl.scala:712)
[info]   at org.apache.spark.sql.hive.client.HiveClientImpl.runSqlHive(HiveClientImpl.scala:702)
[info]   at org.apache.spark.sql.hive.client.VersionsSuite.$anonfun$new$64(VersionsSuite.scala:591)
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 2]
```diff
@@ -733,13 +745,17 @@ private[hive] class HiveClientImpl(
...
           results
 
-        case _ =>
-          if (state.out != null) {
-            // scalastyle:off println
-            state.out.println(tokens(0) + " " + cmd_1)
-            // scalastyle:on println
-          }
-          Seq(proc.run(cmd_1).getResponseCode.toString)
+        case _: SetProcessor | _: ResetProcessor =>
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Hive 1.2.1's CommandProcessors:
image

Hive 3.1.1's CommandProcessors:
image

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 3]
```diff
@@ -856,20 +864,31 @@ private[hive] class HiveClientImpl(
...
   def reset(): Unit = withHiveState {
-    client.getAllTables("default").asScala.foreach { t =>
-      logDebug(s"Deleting table $t")
-      val table = client.getTable("default", t)
+    val allTables = client.getAllTables("default")
+    val (mvs, others) = allTables.asScala.map(t => client.getTable("default", t))
+      .partition(_.getTableType.toString.equals("MATERIALIZED_VIEW"))
+
+    // Remove materialized view first, otherwise caused a violation of foreign key constraint.
+    mvs.foreach { table =>
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Remove materialized view first, otherwise:
```java
Cause: org.apache.derby.shared.common.error.DerbySQLIntegrityConstraintViolationException: DELETE on table 'TBLS' caused a violation of foreign key constraint 'MV_TABLES_USED_FK2' for key (4).  The statement has been rolled back.
at org.apache.derby.impl.jdbc.SQLExceptionFactory.getSQLException(Unknown Source)
at org.apache.derby.impl.jdbc.Util.generateCsSQLException(Unknown Source)
at org.apache.derby.impl.jdbc.TransactionResourceImpl.wrapInSQLException(Unknown Source)
at org.apache.derby.impl.jdbc.TransactionResourceImpl.handleException(Unknown Source)
at org.apache.derby.impl.jdbc.EmbedConnection.handleException(Unknown Source)
at org.apache.derby.impl.jdbc.ConnectionChild.handleException(Unknown Source)
at org.apache.derby.impl.jdbc.EmbedStatement.executeStatement(Unknown Source)
at org.apache.derby.impl.jdbc.EmbedPreparedStatement.executeBatchElement(Unknown Source)
at org.apache.derby.impl.jdbc.EmbedStatement.executeLargeBatch(Unknown Source)
at org.apache.derby.impl.jdbc.EmbedStatement.executeBatch(Unknown Source)
```

**Comment by wangyum** [comment]
cc @dongjoon-hyun

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 4]
```diff
@@ -856,20 +864,31 @@ private[hive] class HiveClientImpl(
   }
 
   def reset(): Unit = withHiveState {
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Thank you for pinging me, @wangyum . 
Shall we make another PR for this `reset` change?

Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: OK.

Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: I believe this `reset` PR can be merged before this `runHive` PR.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 5]
```diff
@@ -584,10 +586,11 @@ class VersionsSuite extends SparkFunSuite with Logging {
 
     test(s"$version: sql read hive materialized view") {
       // HIVE-14249 Since Hive 2.3.0, materialized view is supported.
-      // But skip Hive 3.1 because of SPARK-27074.
-      if (version == "2.3") {
+      if (version == "2.3" || version == "3.1") {
+        val disableRewrite = if (version == "2.3") "" else "DISABLE REWRITE"
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: Could you add a comment about the reason why we use `DISABLE REWRITE` in `Hive 3.x`?

**Comment by wangyum** [comment]
> how can we build against the Hive 3.1 metastore to test? Or do we just have to manually provide hive 3.1 jars for now.

```shell
bin/spark-shell --conf spark.sql.hive.metastore.version=3.1.1 --conf spark.sql.hive.metastore.jars=maven
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala (hunk 5 above)
Comment (wangyum) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: OK.
https://github.com/apache/hive/blob/8fd90ea13e24c2de5e4abbd9e6ea427ede559a6c/ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java#L13456-L13460

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by wangyum** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by wangyum** [comment]
Retest this please.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @wangyum .

Merged to master.
